### PR TITLE
Improve tile error handling

### LINKED ISSUE #2928: Endless tile requests on certain kinds of errors
Certain types of server errors may cause TerriaJS to repeatedly request the same (failing) tiles over and over again, wasting bandwidth and potentially starving other requests. Specifically, if an error occurs while loading an image but then an XMLHttpRequest to the same URL succeeds, TerriaJS will retry the image and repeat this process endlessly. 

Here's an example of a WMTS server that triggers the problem (at least until we address #2927):
https://tiles.openaerialmap.org/5abae68e65bd8f00110f3e42/0/5abae68e65bd8f00110f3e43/wmts

TerriaJS should limit the number of times it is willing to retry image loads. It should also more carefully validate the XMLHttpRequest response. For example, it should set the `Accepts` header and it should verify that the response is actually an image.

### PR DESCRIPTION
**Pull request by kring**
* Fixed a bug that could cause endless tile requests with certain types incorrect server responses.
* Fixed a bug that could cause endless region tile requests when loading a CSV with a time column where none of the column values could actually be interpreted as a time.
* Added automatic retry with jittered, exponential backoff for tile requests that result in a 5xx HTTP status code. This is especially useful for servers that return 503 or 504 under load. Previously, TerriaJS would frequently disable the layer and hit the user with an error message when accessing such servers.

Fixes #2928 (both the WMTS problem and the unrelated CSV problem)
Closes #3083 
Is probably a reasonable workaround for #3092

Sadly I didn't really make this code any more sane. It's still totally crazy.

### CODE REVIEW

**Comment by meh9** [comment]
Can confirm unrelated CSV problem fixed.

**Comment by meh9** [comment]
Can confirm that we are now very much more resistant to 503's at least, tested on AREMI with the following layers plus Australian Topography basemap, and it took a very large number of 503's before the layers started to actually fail. These show the ones that did and did not fail + Australian Topography:

screen shot 2018-10-05 at 10 57 28

I probably got 1000 or so 503's before things started failing, so I'd say we should go ahead and merge.

File: lib/Map/TableStructure.js [hunk 1]
```diff
@@ -451,8 +451,10 @@ function calculateFinishDatesFromStartDates(startJulianDates, localDefaultFinalD
         var n = sortedUniqueJulianDates.length;
         if (n > 1) {
             finalDurationSeconds = JulianDate.secondsDifference(sortedUniqueJulianDates[n - 1], sortedUniqueJulianDates[0]) / (n - 1);
+            endDates.push(JulianDate.addSeconds(sortedUniqueJulianDates[n - 1], finalDurationSeconds, new JulianDate()));
+        } else {
+            endDates.push(undefined);
```
Comment (tephenavies) [inline lib/Map/TableStructure.js:456]: @kring Is this the right thing to do for `n === 1`?

Comment (kring) [reply lib/Map/TableStructure.js:456]: 😬 I think so. The way `endDates` is used below, it looks like there has to be something added, or the indices will be wrong.

File: lib/Models/ImageryLayerCatalogItem.js [hunk 2]
```diff
@@ -586,84 +588,67 @@ ImageryLayerCatalogItem.enableLayer = function(catalogItem, imageryProvider, opa
             }
 
             if (tileProviderError.timesRetried === 0) {
+                // There was an intervening success, so restart our count of the tile failures.
                 tileFailures = 0;
+                tileRetriesByMap = {};
```
Comment (stevage) [inline lib/Models/ImageryLayerCatalogItem.js:603]: retriesPerTile would be a better name.

Comment (stevage) [reply lib/Models/ImageryLayerCatalogItem.js:603]: Also, is it definitely correct to reset it whenever there is any "intervening success"? If tile X has failed 3 times, then some other random tile succeeds, will this (incorrectly) get reset?

File: lib/Models/ImageryLayerCatalogItem.js [hunk 3]
```diff
@@ -705,9 +690,98 @@ ImageryLayerCatalogItem.enableLayer = function(catalogItem, imageryProvider, opa
...
+                        // could not load it as an image.
+                        failTile({
+                            name: 'Tile Error',
+                            message: 'Got a 200 (OK) response from URL:\n' +
+                                     getUrlForImageryTile(imageryProvider, tileProviderError.x, tileProviderError.y, tileProviderError.level) + '\n' +
+                                     'but it could not be loaded as an image.'
+                        });
+                    } else if (!defined(e.statusCode)) {
+                        // On a modern-ish browser, this is almost certainly a CORS problem.
+                        // Note that ignoreUnknownTileErrors is only for genuinely unknown (no status code) errors.
```
Comment (stevage) [inline lib/Models/ImageryLayerCatalogItem.js:778]: This situation also arises on DNS failures (eg, non existent domain)

**Comment by kring** [comment]
@stevage reports that https://tiles.openaerialmap.org/5abae68e65bd8f00110f3e42/0/5abae68e65bd8f00110f3e43/wmts doesn't work on http://ci.terria.io/fail-better/
It does the GetCapabilities successfully, but then doesn't request any tiles for some reason.

**Comment by stevage** [comment]
Actually the last time I looked at the CI, it seemed to be fetching tiles, but then failing silently. I really don't know what's going on. Possibly totally spurious.

**Comment by kring** [comment]
Ok here's what's happening with that openaerialmap server...

The extent of the one layer on that server is a tiny little square near Cape Code, MA, USA. So when you're looking at Australia, Cesium only downloads a couple of tiles. But those tiles all return XML, probably because of #2927, which mentions this exact server. Because we only download a couple of tiles, though, it doesn't fail and show the user a message immediately. But if you try to zoom to the layer's extent, you'll get an error like this pretty quickly:

image

So I think everything is working as expected, but let me know if anything still seems off to you @stevage.

File: lib/Models/ImageryLayerCatalogItem.js (hunk 2 above)
Comment (kring) [reply lib/Models/ImageryLayerCatalogItem.js:603]: It's definitely not _correct_, but the effect should be that TerriaJS is more forgiving of errors, which is closer to a :feature: than a :bug:.

File: lib/Models/ImageryLayerCatalogItem.js (hunk 3 above)
Comment (kring) [reply lib/Models/ImageryLayerCatalogItem.js:778]: Good point!

**Comment by kring** [comment]
I think I've addressed all the review comments @steve9164 and @stevage.

**Comment by stevage** [comment]
>The server denied Cross-Origin Resource Sharing (CORS) access to this URL

More often than not, it will be the browser denying access on the basis that the server didn't provide any CORS instructions at all, right?

**Comment by stevage** [comment]
(I thought I had already pushed these commits, but I pushed them to the wrong branch...)

**Comment by kring** [comment]
> More often than not, it will be the browser denying access on the basis that the server didn't provide any CORS instructions at all, right?

Very true, but I think explaining that just muddies the water. The browser can't be changed, but the server potentially can be, so it's reasonable to explain it as something that has to change on the server. From the browser's perspective, a lack of explicit approval is the same as denial.

**Pull request merged**